pkg/discovery-manager/discovery_controller.go

```diff
@@ -47,13 +56,276 @@ type DiscoveryReconciler struct {
-	_ = log.FromContext(ctx)
+	log := ctrl.LoggerFrom(ctx, "discovery", req.NamespacedName)
+	ctx = ctrl.LoggerInto(ctx, log)
 
-	// TODO(user): your logic here
+	var peeringCandidate *advertisementv1alpha1.PeeringCandidate
+
+	var discovery advertisementv1alpha1.Discovery
+	if err := r.Get(ctx, req.NamespacedName, &discovery); client.IgnoreNotFound(err) != nil {
+		klog.Errorf("Error when getting Discovery %s before reconcile: %s", req.NamespacedName, err)
+		return ctrl.Result{}, err
+	} else if err != nil {
+		klog.Infof("Discovery %s not found, probably deleted", req.NamespacedName)
+		return ctrl.Result{}, nil
+	}
+
+	klog.Infof("Discovery %s started", discovery.Name)
+
+	if discovery.Status.Phase.Phase != nodecorev1alpha1.PhaseSolved &&
+		discovery.Status.Phase.Phase != nodecorev1alpha1.PhaseTimeout &&
+		discovery.Status.Phase.Phase != nodecorev1alpha1.PhaseFailed &&
+		discovery.Status.Phase.Phase != nodecorev1alpha1.PhaseRunning &&
+		discovery.Status.Phase.Phase != nodecorev1alpha1.PhaseIdle {
+
+		setDiscoveryPhase(&discovery, nodecorev1alpha1.PhaseRunning)
+
+		if err := r.updateSolverStatus(ctx, &discovery); err != nil {
+			klog.Errorf("Error when updating Discovery %s status before reconcile: %s", req.NamespacedName, err)
+			return ctrl.Result{}, err
+		}
+		return ctrl.Result{}, nil
+	}
+
+	switch discovery.Status.Phase.Phase {
+	case nodecorev1alpha1.PhaseRunning:
+		flavoursCR, err := getFlavour(discovery.Spec.Selector)
+		if err != nil {
+			klog.Errorf("Error when getting Flavour: %s", err)
+			return ctrl.Result{}, err
+		}
+
+		klog.Infof("Flavours found: %d", len(flavoursCR))
+
+		// first is a boolean that is used to check if the current flavour is the first one in the array
+		var first bool = true
+		var i int = 1
+		for _, flavourCR := range flavoursCR {
+			if first {
+				// We refer to the first peering candidate as the one that is reserved
+				peeringCandidate = createPeeringCandidateCustomResources(flavourCR, &discovery, first, i)
+				klog.Infof("Peering Candidate %s created", peeringCandidate.Name)
+			} else {
+				// the others are created as not reserved
+				peeringCandidateNotReserved := createPeeringCandidateCustomResources(flavourCR, &discovery, first, i)
+				klog.Infof("Peering Candidate %s created but not reserved", peeringCandidateNotReserved.Name)
+				err = r.Create(context.Background(), peeringCandidateNotReserved)
+				if err != nil {
+					klog.Errorf("Error when creating Peering Candidate not reserved: %s", err)
+				}
+			}
+			i++
+			first = false
+		}
+
+		err = r.Create(context.Background(), peeringCandidate)
+		if err != nil {
+			klog.Errorf("Error when creating Peering Candidate: %s", err)
+			setDiscoveryPhase(&discovery, nodecorev1alpha1.PhaseFailed)
+		} else {
+			// Update the Discovery with the PeeringCandidate
+			discovery.Spec.PeeringCandidate.Name = peeringCandidate.Name
+			discovery.Spec.PeeringCandidate.Namespace = DEFAULT_NAMESPACE
```

Comment:
you can create a `GenericRef` obj and than link it inside the `discovery` CR

Reply:
Fixed using a `GenericRef` that links inside the discovery.